### PR TITLE
Replaces plurality voting with approval voting for some votes

### PR DESCRIPTION
**Pull request by Putnam3145**
## About The Pull Request

As the title: replaces the [first past the post](https://en.wikipedia.org/wiki/First-past-the-post_voting) voting method used in the codebase right now with [approval voting](https://en.wikipedia.org/wiki/Approval_voting).

## Why It's Good For The Game

First past the post sucks. I mean, it's just awful. Sometimes the winner is someone who the actual majority of voters doesn't want, it's highly susceptible to strategic voting, it leads to a massive amount of similar perverse results...

Approval voting is simple: you can vote for as many options as you want. For two-choice votes, it's identical, as long as there's one winner; voting both options is the same as voting one option. For multi-winner votes, it's *still* better, because voting both makes the vote closer, which is what you'd *want*.

And if you want to know why approval over IRV: well, [see this article, if you have the time](https://www.electionscience.org/library/approval-voting-versus-irv/). That, plus it's much, much easier to implement approval (see how small this change is!)

## Changelog
:cl:
tweak: Voting now uses approval instead of first-past-the-post
/:cl:

### CODE REVIEW

**Comment by nemvar** [comment]
config would make more sense imo

**Comment by Putnam3145** [comment]
I'll do it if that's the consensus obv. (by which I mean one more person tells me to do it) but I will argue it's not really necessary to do so, since the votes are basically counted the same way and... I really can't emphasize enough how garbage first-past-the-post is.

**Comment by nemvar** [comment]
This system is used for more than just map votes.
It could be used to run a quick "Yes, No, Maybe" poll in which plurality makes no sense.

**Comment by Putnam3145** [comment]
I know it's used for more than just map votes, I pointed out that it's still basically identical for yes/no votes, too.

**Comment by Bokkiewokkie** [comment]
> This system is used for more than just map votes.
> It could be used to run a quick "Yes, No, Maybe" poll in which plurality makes no sense.

This is why I think it should, if it gets added, be an optional thing to add to a poll and not a permanent change to all polls forever.

**Comment by cacogen** [comment]
Bokkiewokkie's right, on custom votes admins should be able to specify how many choices can be selected. Map votes and gamemode votes should allow as many choices as maps, and restart votes should only allow one choice.

**Comment by Putnam3145** [comment]
Alright, yeah, that's all fine. Conveniently, I already have all that code in place from last time I implemented this, but I decided to keep it simple for this PR because that was hellcode (I also implemented: IRV, Schulze, score, majority judgement, and allowed any of them for custom polls. majority judgement and IRV used to work, but don't anymore. I don't know what happened. I don't want to repeat that.)

**Comment by Putnam3145** [comment]
Alright, suggestions implemented.

File: code/__DEFINES/vote.dm [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#define PLURALITY_VOTING 0
+#define APPROVAL_VOTING 1
+
+GLOBAL_LIST_INIT(vote_type_names,list(\
```
Comment (spookydonut) [inline code/__DEFINES/vote.dm]: why is this a global list

Comment (Putnam3145) [reply code/__DEFINES/vote.dm]: Shit, I mindlessly copied that from my Cit implementation where (IIRC) I used it more than once. Yeah, I'll just make it in the actual proc.

**Comment by Putnam3145** [comment]
Yeah, if y'all got anything else I'd be happy to implement it.

**Comment by cacogen** [comment]
pls merge

**Comment by PKPenguin321** [comment]
This is for votes and not polls, yes?

**Comment by Putnam3145** [comment]
Yes, exclusively for in-game votes, started by the "vote" verb or by the server.

**Comment by cacogen** [comment]
still merge. makes map votes less skewed towards meta

File: code/controllers/subsystem/vote.dm [hunk 2]
```diff
@@ -157,22 +158,43 @@ SUBSYSTEM_DEF(vote)
...
+							choices[choices[vote]]--
+						else
+							voted[usr.ckey] += vote
+							choices[choices[vote]]++
+					else
+						voted += usr.ckey
+						voted[usr.ckey] = list(vote)
+						choices[choices[vote]]++
+						return vote
+				else // plurality-by-default
```
Comment (Rohesie) [inline code/controllers/subsystem/vote.dm:176]: `if(PLURALITY_VOTING )`
and
`else CRASH()`

File: code/controllers/subsystem/vote.dm [hunk 3]
```diff
@@ -157,22 +158,43 @@ SUBSYSTEM_DEF(vote)
...
+					if(usr.ckey in voted)
```
Comment (Rohesie) [inline code/controllers/subsystem/vote.dm:177]:
```suggestion
					if(voted[usr.ckey])
```
Faster lookup.

File: code/controllers/subsystem/vote.dm [hunk 4]
```diff
@@ -157,22 +158,43 @@ SUBSYSTEM_DEF(vote)
...
 			return FALSE
-		if(!(usr.ckey in voted))
-			if(vote && 1<=vote && vote<=choices.len)
-				voted += usr.ckey
-				choices[choices[vote]]++	//check this
-				return vote
+		if(vote && ISINRANGE(vote, 1, choices.len))
+			switch(vote_system)
+				if(APPROVAL_VOTING)
+					if(usr.ckey in voted)
```
Comment (Rohesie) [inline code/controllers/subsystem/vote.dm:164]:
```suggestion
					if(voted[usr.ckey])
```

File: code/controllers/subsystem/vote.dm [hunk 5]
```diff
@@ -203,7 +225,10 @@ SUBSYSTEM_DEF(vote)
 			if("custom")
 				question = stripped_input(usr,"What is the vote for?")
 				if(!question)
-					return FALSE
+					return 0
```
Comment (Rohesie) [inline code/controllers/subsystem/vote.dm:228]: Why are you returning zero instead of `FALSE`?

File: code/controllers/subsystem/vote.dm [hunk 6]
```diff
@@ -203,7 +225,10 @@ SUBSYSTEM_DEF(vote)
 			if("custom")
 				question = stripped_input(usr,"What is the vote for?")
 				if(!question)
-					return FALSE
+					return 0
+				var/list/vote_type_names = list("Plurality (can only vote one option)" = PLURALITY_VOTING,"Approval (can vote any amount)" = APPROVAL_VOTING)
+				var/system_string = input(usr,"Which voting type?",vote_type_names[1]) in vote_type_names
```
Comment (Rohesie) [inline code/controllers/subsystem/vote.dm:230]: Aren't you giving here a single choice?

**Comment by ExcessiveUseOfCobblestone** [comment]
20 days to fix code, most of it suggestion blocks, would suggest this is not being updated.

**Comment by cacogen** [comment]
rip

**Pull request closed without merging**